test/e2e/ingress.go

```diff
@@ -0,0 +1,311 @@
+//   - Out of quota (describe ing should show you if this is the case)
+//   - Mismatched service/container port, or endpoints are dead.
+
+var (
+	appPrefix         = "foo-app-"
+	pathPrefix        = "foo"
+	testImage         = "bprashanth/n-way-http:1.0"
+	httpContainerPort = 8080
+
+	// Average duration of loadbalancer startup.
+	avgLBLatency = 7 * time.Minute
+
+	// On average it takes ~6 minutes for a single backend to come online.
+	// We *don't* expect this poll to consistently take 15 minutes for every
+	// Ingress as GCE is creating/checking backends in parallel, but at the
+	// same time, we're not testing GCE startup latency. So give it enough
+	// time, and fail if the average is too high.
+	lbPollTimeout  = 15 * time.Minute
+	lbPollInterval = 30 * time.Second
+
+	// One can scale this test by tweaking numApps and numIng, the former will
+	// create more RCs/Services and add them to a single Ingress, while the latter
+	// will create smaller, more fragments Ingresses.
+	numApps = 6
+	numIng  = 3
+)
+
+// timeSlice allows sorting of time.Duration
+type timeSlice []time.Duration
+
+func (p timeSlice) Len() int {
+	return len(p)
+}
+
+func (p timeSlice) Less(i, j int) bool {
+	return p[i] < p[j]
+}
+
+func (p timeSlice) Swap(i, j int) {
+	p[i], p[j] = p[j], p[i]
+}
+
+// ruleByIndex returns an IngressRule for the given index.
+func ruleByIndex(i int) extensions.IngressRule {
+	return extensions.IngressRule{
+		Host: fmt.Sprintf("foo%d.bar.com", i),
+		IngressRuleValue: extensions.IngressRuleValue{
+			HTTP: &extensions.HTTPIngressRuleValue{
+				Paths: []extensions.HTTPIngressPath{
+					{
+						Path: fmt.Sprintf("/%v%d", pathPrefix, i),
+						Backend: extensions.IngressBackend{
+							ServiceName: fmt.Sprintf("%v%d", appPrefix, i),
+							ServicePort: util.NewIntOrStringFromInt(httpContainerPort),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+// createIngress creates an Ingress with num rules. Eg:
+// start = 1 num = 2 will given you a single Ingress with 2 rules:
+// Ingress {
+//	 foo1.bar.com: /foo1
+//	 foo2.bar.com: /foo2
+// }
+func createIngress(c *client.Client, ns string, start, num int) extensions.Ingress {
+	ing := extensions.Ingress{
+		ObjectMeta: api.ObjectMeta{
+			Name:      fmt.Sprintf("%v%d", appPrefix, start),
+			Namespace: ns,
+		},
+		Spec: extensions.IngressSpec{
+			Backend: &extensions.IngressBackend{
+				ServiceName: fmt.Sprintf("%v%d", appPrefix, start),
+				ServicePort: util.NewIntOrStringFromInt(httpContainerPort),
+			},
+			Rules: []extensions.IngressRule{},
+		},
+	}
+	for i := start; i < start+num; i++ {
+		ing.Spec.Rules = append(ing.Spec.Rules, ruleByIndex(i))
+	}
+	Logf("Creating ingress %v", start)
+	_, err := c.Extensions().Ingress(ns).Create(&ing)
+	Expect(err).NotTo(HaveOccurred())
+	return ing
+}
+
+// createApp will create a single RC and Svc. The Svc will match pods of the
+// RC using the selector: 'name'=<name arg>
+func createApp(c *client.Client, ns string, i int) {
```

Comment:
nit: This seems like a very useful utility for more than just this test.  Suggest factoring it out into e2e/util library.  In the process you'll want to improve its decoupling by passing in appPrefix, httpContainerPort, testImage etc as parameters.
Happy to do all this in a follow-on PR if you prefer.  Just create an issue to track that.